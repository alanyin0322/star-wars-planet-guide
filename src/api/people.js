import axios from 'axios';

export async function fetchPlanetPeopleName(residents) {
  try{
    let residentNames = [];
    for (let i = 0; i < residents.length; i++){
      const { data: {name} } = await axios.get(residents[i]);
      residentNames = residentNames.concat( name );
    }
    return residentNames
  } catch(err) {
    alert('Error fetching people on the planet');
  }
}