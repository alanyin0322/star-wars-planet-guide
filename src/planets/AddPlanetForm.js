import React from 'react';

export default class AddPlanetForm extends React.Component {
  state = {
    name: '',
    population: '',
    terrain: '',
    error: '',
    success: ''
  };

  onNameChange = e => {
    this.setState({name: e.target.value});
  };


  onPopulationChange = e => {
    const population = e.target.value;
  if (population.match(/^\d*$/)){
      this.setState({population});
    }
  };

  onTerrainChange = e => {
    this.setState({terrain: e.target.value});
  };

  onSubmit = e => {
    e.preventDefault();
    if(!this.state.name || !this.state.population || !this.state.terrain){
      this.setState({error: 'All fields are required.'});
      this.setState({success: ''});
    } else {
      const planetData = {
        name: this.state.name,
        population: this.state.population,
        terrain: this.state.terrain
      };
      this.props.addPlanet(planetData);
      this.setState({
        name: '',
        population: '',
        terrain: '',
        error:'',
        success: 'New planet added!'
      });
    }
  }

  render(){
    return (
        <thead>
          <tr className="table-active">
            <th>
            </th>
            <th>
              <input type="text" placeholder="e.g. Cantonica" value={this.state.name} onChange={this.onNameChange}/>
            </th>
            <th>
              <input type="text" placeholder="100" value={this.state.population} onChange={this.onPopulationChange}/>
            </th>
            <th>
              <input type="text" placeholder="e.g. ocean" value={this.state.terrain} onChange={this.onTerrainChange}/>
            </th>
            <th>
              <button onClick={this.onSubmit} className="btn btn-add">Add Planet</button>
            </th>
          </tr>
          {this.state.error ? (
            <tr className="table-active">
              <td></td>
              <td><p style={{color: 'red'}}>{this.state.error}</p></td>
              <td></td>
              <td></td>
              <td></td>
            </tr>
          ) : null}
          {this.state.success ? (
            <tr className="table-active">
              <td></td>
              <td><p style={{color: 'red'}}>{this.state.success}</p></td>
              <td></td>
              <td></td>
              <td></td>
            </tr>
          ) : null}
        </thead>
    );
  }
}