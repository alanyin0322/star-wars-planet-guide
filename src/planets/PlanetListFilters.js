import React from 'react';

export default class PlanetListFilters extends React.Component {

  render() {
    return (
      <div>
        <input type="text" value={this.props.filter} onChange={this.props.setFilter}/>
        <select value={this.props.sortBy} onChange={this.props.setSortBy} >
          <option value="name-a-to-z">Name: A to Z</option>
          <option value="name-z-to-a">Name: Z to A</option>
          <option value="population-high-to-low">Population: High to Low</option>
          <option value="population-low-to-high">Population: Low to High</option>
        </select>
      </div>
    );
  }
} 