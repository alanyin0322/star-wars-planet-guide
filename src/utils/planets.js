export const getIdFromUrl = url => {
  const urlFragments = url.split('/');
  return urlFragments[urlFragments.length - 2];
};

export const getVisiblePlanets = ({ planets, filter, sortBy }) => {
  return planets.filter(planet => {
    const filterMatch = typeof filter !== 'string' || 
    planet.name.toLowerCase().includes(filter.toLowerCase());
    return filterMatch;
  }).sort((a, b) => {
    if (sortBy === 'name-a-to-z') {
      return a.name > b.name ? 1 : -1;
    } else if (sortBy === 'name-z-to-a') {
      return a.name < b.name ? 1 : -1;
    } else if (sortBy === 'population-low-to-high') {
      return a.population > b.population ? 1 : -1;
    } else if (sortBy === 'population-high-to-low') {
      return a.population < b.population ? 1 : -1;
    }
  });
};