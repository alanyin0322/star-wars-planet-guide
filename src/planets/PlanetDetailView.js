import React from 'react';
import { Link } from 'react-router-dom';

import LoadingSpinner from '../UIs/LoadingSpinner';

import { fetchPlanetById } from '../api/planets';


export default class PlanetDetailView extends React.Component {
  state = {
    isLoading: false,
    planet: {}
  };

  componentDidMount() {
    this.setState({isLoading: true});
    fetchPlanetById(this.props.match.params.id)
      .then(planet => {
        this.setState ({
          isLoading: false,
          planet
        });
      })
      .catch(e => {
        alert('error fetching planet details');
      });
  }

  render() {
    const { 
      name = '',
      rotPeriod = '',
      orbPeriod = '',
      diameter = '',
      climate = '',
      gravity = '',
      terrain = '',
      surfaceWater = '',
      population = '',
      residentNames = []
    } = this.state.planet;

    const cardStyle = {width: '50rem'};

    return this.state.isLoading ? (
      <div className="planet-card">
        <LoadingSpinner 
          color="#FCE852" 
          className="planet-card__loader" 
          size={100}
        />
      </div>
    ) : (
      <div className="container">
        <div className="card text-center planet-card" style={cardStyle}>
          <img 
            className="card-img-top" src="https://drscdn.500px.org/photo/14992149/q%3D80_m%3D2000/v2?webp=true&sig=3076023a9a09be6772d42759668d8498ef224a117b731d04c54634e2fd468ce1" 
            alt="Planet"
          />
          <h2 className="card-header">
            {name}
          </h2>
          <ul className="list-group list-group-flush">
            <li className="list-group-item">Diameter: {diameter}</li>
            <li className="list-group-item">Population: {population}</li>
            <li className="list-group-item">Gravity: {gravity}</li>
            <li className="list-group-item">Terrain: {terrain}</li>
            <li className="list-group-item">Climate: {climate}</li>
            <li className="list-group-item">Surface Water: {surfaceWater}</li>
            <li className="list-group-item">Orbital Period: {orbPeriod}</li>
            <li className="list-group-item">Rotation Period: {rotPeriod}</li>
            <li className="list-group-item">Residents: 
              <ul className="planet-card__list">
                {residentNames.map(residentName => (
                  <li key={residentName}>{residentName}</li>
                ))}
              </ul>
              <Link className='btn btn-primary' to={`/`}>
                Back
              </Link>
            </li>

          </ul>
        </div>
      </div>
    );
  }
}

