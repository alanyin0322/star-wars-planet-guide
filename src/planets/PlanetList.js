import React from 'react';
import PlanetListItem from './PlanetListItem';
import AddPlanetForm from './AddPlanetForm';

const PlanetList = (props) => (
  <div>
    {
      props.planets.length === 0 ? (
        <p>No matching planet</p>
      ) : (
        <div className='container'>
          <table className='table'>
            <thead>
              <tr>
                <th>#</th>
                <th>Name</th>
                <th>Population</th>
                <th>Terrain</th>
                <th></th>
              </tr>
            </thead>
            <AddPlanetForm addPlanet={props.addPlanet}/>
            <tbody>
              {
                props.planets.map((planet, index) => (
                  <PlanetListItem key={planet.name} {...planet} index={index}/>
                ))
              }
            </tbody>
          </table>
        </div>
      )
    }
  </div>
);

export default PlanetList;