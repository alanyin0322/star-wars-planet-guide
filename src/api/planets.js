import axios from 'axios';
import { getIdFromUrl } from '../utils/planets';

export async function fetchPlanets() {
  try{
    const { data: { count } } = await axios.get('/planets');
    const pageNum = Math.ceil(count / 10);

    let allResults = [];
    for (let i = 1; i <= pageNum; i++){
      const { data: { results } } = await axios.get(`/planets?page=${i}`);
      allResults = allResults.concat(results);
    }

    return allResults.map(result => ({
      ...result,
      id: getIdFromUrl(result.url)
    }))
  } catch(err) {
    console.log(err);
  }
}

export async function fetchPlanetById(Id) {
  const url = `/planets/${Id}`;
  try{
    const { data } = await axios.get(url);
    const { name, rotation_period: rotPeriod, orbital_period: orbPeriod, diameter, climate, gravity, terrain, surface_water: surfaceWater, population, residents } = data; 
    return {
      name, 
      rotPeriod,
      orbPeriod,
      diameter,
      climate, 
      gravity,
      terrain,
      surfaceWater,
      population,
      residents
    }
  } catch(err) {
    console.log(err);
  }
}