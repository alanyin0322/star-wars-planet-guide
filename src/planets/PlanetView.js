import React from 'react';

import LoadingPage from '../components/LoadingPage';
import PlanetList from './PlanetList';
import PlanetListFilters from './PlanetListFilters';

import { getPlanets } from '../api/planets';
import { getVisiblePlanets } from '../utils/planets';

export default class PlanetView extends React.Component {
  state = {
    isLoading: true,
    planets: [],
    filter: '',
    sortBy: 'name'
  };

  componentDidMount() {
    this.setState({isLoading: true});
    getPlanets()
      .then((planetResults) => {
        this.setState({
          isLoading: false,
          planets: planetResults
        });
      });
  }

  setFilter = e => {
    console.log('filter changing');
    this.setState({filter: e.target.value});
  }

  setSortBy = e => {
    console.log('sortby changing');
    this.setState({sortBy: e.target.value});
  }

  render() {
    const visiblePlanets = getVisiblePlanets(this.state);
    return this.state.isLoading ? (
      <LoadingPage />
    ) : (
      <div className="container">
        <div className="row d-flex flex-row align-items-center">
          <div className="col-3">
            <p>Displaying {visiblePlanets.length} of {this.state.planets.length} planets</p>
          </div>
          <div className="col-9">
            <PlanetListFilters 
              setFilter={this.setFilter} 
              setSortBy={this.setSortBy} 
              filter={this.state.filter} 
              sortBy={this.state.sortBy} 
            />
          </div>
        </div>
        <PlanetList planets={visiblePlanets} />
      </div>
    );
  }
}