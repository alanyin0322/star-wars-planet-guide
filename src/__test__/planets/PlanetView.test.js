import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import PlanetView from '../../planets/PlanetView';
import * as api from '../../api/planets';
import planetResults from '../fixtures/planets';
import PlanetListFilters from '../../planets/PlanetListFilters';

describe('PlanetView component',() => {
  test('PlanetView component matches snapshot', () => {
    const tree = shallow(<PlanetView />);
    tree.setState({isLoading: false});
    expect(toJson(tree)).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    test('componentDidMount gets called', () => {
      const spy = jest.spyOn(PlanetView.prototype, 'componentDidMount');
      const wrapper = shallow(<PlanetView />);
      expect(spy).toHaveBeenCalledTimes(1);
    });

    test('componentDidMount makes api call and retrieves data', async () => {
      const spy = jest.spyOn(api, 'getPlanets').mockImplementation(() => Promise.resolve(planetResults));
      const wrapper = await shallow(<PlanetView />);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(wrapper.instance().state.planets.length).toBe(3);
    });
  });

  describe('setFilter method', () => {
    test('correctly gets passed to PlanetListFilter component', () => {
      const wrapper = shallow(<PlanetView />);
      wrapper.setState({isLoading: false});
      expect(wrapper.find(PlanetListFilters).prop('setFilter')).toEqual(wrapper.instance().setFilter);
    });

    test('correctly set the state when gets called', () => {
      const wrapper = shallow(<PlanetView />);
      wrapper.instance().setFilter({target: {value: 'a'}});
      expect(wrapper.state().filter).toBe('a');
    });
  });

  describe('setSortBy method', () => {
    test('correctly gets passed to PlanetListFilter component', () => {
      const wrapper = shallow(<PlanetView />);
      wrapper.setState({isLoading: false});
      expect(wrapper.find(PlanetListFilters).prop('setSortBy')).toEqual(wrapper.instance().setSortBy);
    });

    test('correctly set the state when gets called', () => {
      const wrapper = shallow(<PlanetView />);
      wrapper.instance().setSortBy({target: {value: 'population-high-to-low'}});
      expect(wrapper.state().sortBy).toBe('population-high-to-low');
    });
  });
});
