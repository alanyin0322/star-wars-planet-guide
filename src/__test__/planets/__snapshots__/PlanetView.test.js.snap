// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetView component PlanetView component matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row d-flex flex-row align-items-center"
  >
    <div
      className="col-3"
    >
      <p>
        Displaying 
        0
         of 
        0
         planets
      </p>
    </div>
    <div
      className="col-9"
    >
      <PlanetListFilters
        filter=""
        setFilter={[Function]}
        setSortBy={[Function]}
        sortBy="name-a-to-z"
      />
    </div>
  </div>
  <PlanetList
    planets={Array []}
  />
</div>
`;
